ql/src/java/org/apache/hadoop/hive/ql/optimizer/SemiJoinReductionMerge.java

```diff
@@ -0,0 +1,399 @@
+    params.remove(params.size() - 1);
+    return params;
+  }
+
+  private static List<ExprNodeDesc> createGroupByAggregationParameters(ReduceSinkOperator reduceOp) {
+    List<ExprNodeDesc> params = new ArrayList<>();
+    // There is a 1-1 mapping between columns and parameters for the aggregation functions min, max, bloom
+    for (ColumnInfo c : reduceOp.getSchema().getSignature()) {
+      String name = Utilities.ReduceField.VALUE + "." + c.getInternalName();
+      params.add(new ExprNodeColumnDesc(new ColumnInfo(name, c.getType(), "", false)));
+    }
+    return params;
+  }
+
+  private static ExprNodeGenericFuncDesc and(List<ExprNodeDesc> args) {
+    return new ExprNodeGenericFuncDesc(TypeInfoFactory.booleanTypeInfo, new GenericUDFOPAnd(), "and", args);
+  }
+
+  private static AggregationDesc minAggregation(GenericUDAFEvaluator.Mode mode, ExprNodeDesc col) {
+    List<ExprNodeDesc> p = Collections.singletonList(col);
+    return new AggregationDesc("min", new GenericUDAFMin.GenericUDAFMinEvaluator(), p, false, mode);
+  }
+
+  private static AggregationDesc maxAggregation(GenericUDAFEvaluator.Mode mode, ExprNodeDesc col) {
+    List<ExprNodeDesc> p = Collections.singletonList(col);
+    return new AggregationDesc("max", new GenericUDAFMax.GenericUDAFMaxEvaluator(), p, false, mode);
+  }
+
+  private static AggregationDesc bloomFilterAggregation(GenericUDAFEvaluator.Mode mode, ExprNodeDesc col,
+      SelectOperator source, HiveConf conf) {
+    GenericUDAFBloomFilter.GenericUDAFBloomFilterEvaluator bloomFilterEval =
+        new GenericUDAFBloomFilter.GenericUDAFBloomFilterEvaluator();
+    bloomFilterEval.setSourceOperator(source);
+    bloomFilterEval.setMaxEntries(conf.getLongVar(HiveConf.ConfVars.TEZ_MAX_BLOOM_FILTER_ENTRIES));
+    bloomFilterEval.setMinEntries(conf.getLongVar(HiveConf.ConfVars.TEZ_MIN_BLOOM_FILTER_ENTRIES));
+    bloomFilterEval.setFactor(conf.getFloatVar(HiveConf.ConfVars.TEZ_BLOOM_FILTER_FACTOR));
+    // TODO Setup hints
+    List<ExprNodeDesc> p = Collections.singletonList(col);
+    AggregationDesc bloom = new AggregationDesc("bloom_filter", bloomFilterEval, p, false, mode);
+    // TODO Why do we need to set it explicitly?
+    bloom.setGenericUDAFWritableEvaluator(bloomFilterEval);
+    return bloom;
+  }
+
+  private static final class SJSourceTarget {
+    private final Operator<?> source;
+    private final TableScanOperator target;
+
+    public SJSourceTarget(Operator<?> source, TableScanOperator target) {
+      this.source = source;
+      this.target = target;
+    }
+
+    @Override public boolean equals(Object o) {
+      if (this == o)
+        return true;
+      if (o == null || getClass() != o.getClass())
+        return false;
+
+      SJSourceTarget that = (SJSourceTarget) o;
+
+      if (!source.equals(that.source))
+        return false;
+      return target.equals(that.target);
+    }
+
+    @Override public int hashCode() {
+      int result = source.hashCode();
```

Comment:
I agree that the proposed change makes the code more readable, but it is also slightly less efficient since we are instantiating an Object[] array only to throw it away later on.

Reply:
Fair enough... Though I would be surprised if such overhead would be noticeable in our query compilation latency :)